jib-core/src/test/java/com/google/cloud/tools/jib/tar/TarExtractorTest.java

```diff
@@ -79,4 +84,52 @@ public void testExtract_symlinks() throws URISyntaxException, IOException {
     Assert.assertTrue(Files.isSymbolicLink(destination.resolve("directory-symlink")));
     Assert.assertTrue(Files.isSymbolicLink(destination.resolve("directory1/file-symlink")));
   }
+
+  @Test
+  public void testExtract_modificationTimePreserved() throws URISyntaxException, IOException {
+    Path source = Paths.get(Resources.getResource("core/extract.tar").toURI());
+    Path destination = temporaryFolder.getRoot().toPath();
+
+    TarExtractor.extract(source, destination);
+
+    assertThat(Files.getLastModifiedTime(destination.resolve("file A")))
+        .isEqualTo(FileTime.from(Instant.parse("2019-08-01T16:13:09Z")));
+    assertThat(Files.getLastModifiedTime(destination.resolve("file B")))
+        .isEqualTo(FileTime.from(Instant.parse("2019-08-01T16:12:00Z")));
+    assertThat(Files.getLastModifiedTime(destination.resolve("folder")))
+        .isEqualTo(FileTime.from(Instant.parse("2019-08-01T16:12:33Z")));
+    assertThat(Files.getLastModifiedTime(destination.resolve("folder/nested folder")))
+        .isEqualTo(FileTime.from(Instant.parse("2019-08-01T16:13:30Z")));
+    assertThat(Files.getLastModifiedTime(destination.resolve("folder/nested folder/file C")))
+        .isEqualTo(FileTime.from(Instant.parse("2019-08-01T16:12:21Z")));
+  }
+
+  @Test
+  public void testExtract_reproducibleTimestampsEnabled() throws URISyntaxException, IOException {
+    // The tarfile has only level1/level2/level3/file.txt packaged
+    Path source = Paths.get(Resources.getResource("core/tarfile-only-file-packaged.tar").toURI());
+
+    Path destination = temporaryFolder.getRoot().toPath();
+
+    TarExtractor.extract(source, destination, true);
+
+    assertThat(Files.getLastModifiedTime(destination.resolve("level-1")))
+        .isEqualTo(FileTime.fromMillis(1000L));
+    assertThat(Files.getLastModifiedTime(destination.resolve("level-1/level-2")))
+        .isEqualTo(FileTime.fromMillis(1000L));
+    assertThat(Files.getLastModifiedTime(destination.resolve("level-1/level-2/level-3")))
+        .isEqualTo(FileTime.fromMillis(1000L));
+    assertThat(Files.getLastModifiedTime(destination.resolve("level-1/level-2/level-3/file.txt")))
+        .isEqualTo(FileTime.from(Instant.parse("2021-01-29T21:10:02Z")));
+  }
+
+  @Test
+  public void testExtract_reproducibleTimestampsEnabled_invalidDestination() throws IOException {
+    Path destination = temporaryFolder.getRoot().toPath();
+    Files.createDirectory(destination.resolve("directory"));
+
+    assertThrows(
+        IllegalStateException.class,
+        () -> TarExtractor.extract(Paths.get("ignore"), destination, true));
+  }
```

Comment:
Let's assert the exception message too, since anyone can throw the very common `IllegalStateException`.

Reply:
Good point! Since it's a common exception that makes sense. I tend to err on the side of caution when it comes to asserting the full message of the exception since it makes the test brittle so I tested just the relevant part.